Client/Client.cc

```diff
@@ -504,16 +523,57 @@ Cluster::~Cluster()
 std::pair<uint64_t, Configuration>
 Cluster::getConfiguration() const
 {
-    return clientImpl->getConfiguration();
+    GetConfigurationResult r = getConfiguration2();
+    return std::make_pair(r.configuration, r.servers);
+}
+
+GetConfigurationResult
+Cluster::getConfiguration2(uint64_t timeoutNanoseconds) const
+{
+    return clientImpl->getConfiguration(
+        ClientImpl::absTimeout(timeoutNanoseconds));
+}
+
+GetConfigurationResult
+Cluster::getConfiguration2Ex(uint64_t timeoutNanoseconds) const
+{
+    GetConfigurationResult result(getConfiguration2(timeoutNanoseconds));
+    if (result.status == GetConfigurationResult::TIMEOUT) {
+        throw TimeoutException(result.error);
+    }
+    return result;
 }
 
 ConfigurationResult
 Cluster::setConfiguration(uint64_t oldId,
-                          const Configuration& newConfiguration)
+                          const Configuration& newConfiguration,
+                          uint64_t timeoutNanoseconds)
 {
-    return clientImpl->setConfiguration(oldId, newConfiguration);
+    return clientImpl->setConfiguration(
+        oldId, newConfiguration, ClientImpl::absTimeout(timeoutNanoseconds));
 }
 
+ConfigurationResult
+Cluster::setConfigurationEx(uint64_t oldId,
+                            const Configuration& newConfiguration,
+                            uint64_t timeoutNanoseconds)
+{
+    ConfigurationResult result = setConfiguration(oldId,
+                                                  newConfiguration,
+                                                  timeoutNanoseconds);
+    if (result.status == ConfigurationResult::BAD)
+    {
+        throw ConfigurationException(result.error);
+    }
+    if (result.status == ConfigurationResult::CHANGED)
+    {
+        throw ConfigurationException(result.error);
```

Comment:
So this is what ConfigurationException is for, eh? Seems like one exception is not enough, should be ConfigurationBadException and ConfigurationChangedException, since clients may want to distinguish these.